File: electron/database/appSettingsRepository.ts
import type Database from "better-sqlite3";

export function createAppSettingsRepository(db: Database.Database) {
  const stmts = {
    get: db.prepare(`SELECT value FROM app_settings WHERE key = ?`),
    set: db.prepare(`
      INSERT INTO app_settings (key, value, updated_at)
      VALUES (?, ?, datetime('now'))
      ON CONFLICT(key) DO UPDATE SET value = excluded.value, updated_at = datetime('now')
    `),
    getAll: db.prepare(`SELECT key, value FROM app_settings`),
    remove: db.prepare(`DELETE FROM app_settings WHERE key = ?`),
  };

  return {
    get(key: string): string | null {
      const row = stmts.get.get(key) as { value: string } | undefined;
      return row?.value ?? null;
    },

    set(key: string, value: string): void {
      stmts.set.run(key, value);
    },

    getAll(): Record<string, string> {
      const rows = stmts.getAll.all() as { key: string; value: string }[];
      const result: Record<string, string> = {};
      for (const row of rows) {
        result[row.key] = row.value;
      }
      return result;
    },

    remove(key: string): void {
      stmts.remove.run(key);
    },
  };
}

export type AppSettingsRepository = ReturnType<
  typeof createAppSettingsRepository
>;

File: src-tauri/src/db/app_settings_repository.rs
use rusqlite::{params, Connection};
use std::collections::HashMap;

pub fn get(conn: &Connection, key: &str) -> rusqlite::Result<Option<String>> {
    let mut stmt = conn.prepare("SELECT value FROM app_settings WHERE key = ?1")?;
    let result = stmt.query_row([key], |row| row.get::<_, String>(0));
    match result {
        Ok(value) => Ok(Some(value)),
        Err(rusqlite::Error::QueryReturnedNoRows) => Ok(None),
        Err(e) => Err(e),
    }
}

pub fn set(conn: &Connection, key: &str, value: &str) -> rusqlite::Result<()> {
    conn.execute(
        "INSERT INTO app_settings (key, value, updated_at) VALUES (?1, ?2, datetime('now')) \
         ON CONFLICT(key) DO UPDATE SET value = excluded.value, updated_at = excluded.updated_at",
        params![key, value],
    )?;
    Ok(())
}

pub fn get_all(conn: &Connection) -> rusqlite::Result<HashMap<String, String>> {
    let mut stmt = conn.prepare("SELECT key, value FROM app_settings")?;
    let rows = stmt.query_map([], |row| {
        Ok((row.get::<_, String>(0)?, row.get::<_, String>(1)?))
    })?;

    let mut map = HashMap::new();
    for row in rows {
        let (k, v) = row?;
        map.insert(k, v);
    }
    Ok(map)
}

pub fn remove(conn: &Connection, key: &str) -> rusqlite::Result<()> {
    conn.execute("DELETE FROM app_settings WHERE key = ?1", [key])?;
    Ok(())
}
